### ISSUE TITLE
Async CatalogProviderList and CatalogProvider trait

### ISSUE THREAD

**Issue by lewiszlw** [apache/arrow-datafusion#9918]
### Is your feature request related to a problem or challenge?

My case is to retrive catalog/schema/table through network. Why trait  `CatalogProviderList` and `CatalogProvider` are not async?

**Comment by edmondop** [comment]
@lewiszlw this relates to other type of work I need to do in other projects, can I have a look?

**Comment by lewiszlw** [comment]
> @lewiszlw this relates to other type of work I need to do in other projects, can I have a look?

That is our company project, it's closed source.

**Comment by tshauck** [comment]
FWIW I've wondered this as well. My use case is using a database to store catalog/schema/table data, and right now I just make blocking calls, but fetching catalog data does seems like a common use case.

**Comment by lewiszlw** [comment]
> FWIW I've wondered this as well. My use case is using a database to store catalog/schema/table data, and right now I just make blocking calls, but fetching catalog data does seems like a common use case.

Yeah. Maybe it's not a common use case, but making trait async won't hurt normal case I think.
